### PR TITLE
CDAP-13558 Support CDC pipelines on Spark version 2.x

### PR DESCRIPTION
**Pull request by mlozbin-cybervisiontech**
JIRA - https://issues.cask.co/browse/CDAP-13558
WIKI - https://wiki.cask.co/display/CE/CDC+Solution+Enhancements

In scope of this PR:
- implemented integration tests for Bigtable Sink plugin

### CODE REVIEW

**Comment by mlozbin-cybervisiontech** [comment]
@albertshau, could you please take a look.

File: src/test/java/co/cask/cdc/plugins/integration/sink/CDCBigTablePluginTest.java [hunk 1]
```diff
@@ -0,0 +1,387 @@
...
+
+    @Before
+    public void beforeMethod() {
+      Assume.assumeNotNull(PROJECT);
+      Assume.assumeNotNull(INSTANCE);
+    }
+
+    @Test
+    public void testDeploymentFailedWhenCredentialsFileIsMissing() throws Exception {
+      ETLPlugin sourceConfig = MockSource.getPlugin(Schemas.CHANGE_SCHEMA, Collections.emptyList());
```
Comment (albertshau) [inline src/test/java/co/cask/cdc/plugins/integration/sink/CDCBigTablePluginTest.java]: There is no need to test this through pipeline deployment. It is much less expensive to test these things in a separate CDCBigTableConfigTest class where you create instances of CDCBigTableConfig and just call validate().

File: src/test/java/co/cask/cdc/plugins/integration/sink/CDCBigTablePluginTest.java [hunk 2]
```diff
@@ -0,0 +1,387 @@
...
+    return result.getValue(Bytes.toBytes(CDCTableUtil.CDC_COLUMN_FAMILY), Bytes.toBytes(column));
+  }
+
+  public static class Validations extends CDCPluginTestBase {
+    @Rule
+    public TestName testName = new TestName();
+
+    @Before
+    public void beforeMethod() {
+      Assume.assumeNotNull(PROJECT);
```
Comment (albertshau) [inline src/test/java/co/cask/cdc/plugins/integration/sink/CDCBigTablePluginTest.java]: these tests are testing validation and should always run, regardless of whether or not an actual project or instance is being provided.

File: src/test/java/co/cask/cdc/plugins/integration/sink/CDCBigTablePluginTest.java [hunk 3]
```diff
@@ -0,0 +1,387 @@
...
+        .collect(Collectors.toList());
+
+      ETLPlugin sourceConfig = MockSource.getPlugin(Schemas.CHANGE_SCHEMA, input);
+
+      programManager = deployETL(sourceConfig, sinkConfig, APP_NAME);
+      programManager.startAndWaitForRun(ProgramRunStatus.RUNNING, 10, TimeUnit.SECONDS);
+
+      Awaitility.await().atMost(TEST_TIMEOUT).pollInterval(Duration.ONE_SECOND)
+        .ignoreException(StatusRuntimeException.class).untilAsserted(() -> {
+        TableName tableName = TableName.valueOf(dbTableName);
```
Comment (albertshau) [inline src/test/java/co/cask/cdc/plugins/integration/sink/CDCBigTablePluginTest.java]: nit: everything in the block should be indented

File: src/test/java/co/cask/cdc/plugins/integration/sink/CDCBigTablePluginTest.java [hunk 4]
```diff
@@ -0,0 +1,387 @@
...
```
Comment (albertshau) [inline src/test/java/co/cask/cdc/plugins/integration/sink/CDCBigTablePluginTest.java]: nit: everything in the block should be indented

File: src/test/java/co/cask/cdc/plugins/integration/sink/CDCBigTablePluginTest.java [hunk 5]
```diff
@@ -0,0 +1,387 @@
...
+
+import java.io.IOException;
+import java.util.Collections;
+import java.util.List;
+import java.util.Map;
+import java.util.concurrent.TimeUnit;
+import java.util.stream.Collectors;
+import java.util.stream.Stream;
+
+@RunWith(Enclosed.class)
```
Comment (albertshau) [inline src/test/java/co/cask/cdc/plugins/integration/sink/CDCBigTablePluginTest.java]: I'm not familiar with this, what is the benefit of having inner test classes instead of multiple top level test classes?

File: src/test/java/co/cask/cdc/plugins/integration/sink/CDCBigTablePluginTest.java (hunk 1 above)
Comment (mlozbin-cybervisiontech) [reply src/test/java/co/cask/cdc/plugins/integration/sink/CDCBigTablePluginTest.java]: I did it for several reasons:
- to make tests look similar because we will still need config tests with real pipeline. Splitting those tests will make it more difficult to understand what exactly is covered by tests.
- to make future refactoring easier. As for me "validate" methods inside configuration DTO object violates single responsibility principle a bit.
- to test mappings (if everything is correctly mapped from properties map to config instance)

What do you think?

File: src/test/java/co/cask/cdc/plugins/integration/sink/CDCBigTablePluginTest.java (hunk 5 above)
Comment (mlozbin-cybervisiontech) [reply src/test/java/co/cask/cdc/plugins/integration/sink/CDCBigTablePluginTest.java]: in general it allows you to encapsulate tests with different preconditions (before/after) and some common logic/variables in one class (easier to describe clear testing conventions - one functionality - one test class).

File: src/test/java/co/cask/cdc/plugins/integration/sink/CDCBigTablePluginTest.java (hunk 1 above)
Comment (albertshau) [reply src/test/java/co/cask/cdc/plugins/integration/sink/CDCBigTablePluginTest.java]: The only plugin specific logic that is being tested here is the validation logic in the config. Everything else is functionality that CDAP is supposed to provide, and is unit tested in the CDAP project. When possible, we just want to test plugin logic, not to test plugin logic while also testing all the stuff CDAP does.

**Comment by mlozbin-cybervisiontech** [comment]
@albertshau, addressed all comments.

File: src/main/java/co/cask/cdc/plugins/common/CDCReferencePluginConfig.java [hunk 6]
```diff
@@ -0,0 +1,22 @@
...
+ * {@link ReferencePluginConfig} with reference name validation.
+ */
+public class CDCReferencePluginConfig extends ReferencePluginConfig {
+  public CDCReferencePluginConfig(String referenceName) {
+    super(referenceName);
+  }
+
+  public void validate() {
+    if (!containsMacro(Constants.Reference.REFERENCE_NAME) && !References.isValid(referenceName)) {
+      String message = String.format("%s is not a valid id. Allowed characters are letters, numbers, " +
```
Comment (albertshau) [inline src/main/java/co/cask/cdc/plugins/common/CDCReferencePluginConfig.java]: This message fits better in the References class, as that is where the Pattern is. I think it would be better if References.isValid() was instead References.validate() that throws an exception. This method would then catch that exception and use it's message for the InvalidConfigPropertyException.

File: src/main/java/co/cask/cdc/plugins/common/References.java [hunk 7]
```diff
@@ -0,0 +1,34 @@
+/*
+ * Copyright © 2016 Cask Data, Inc.
```
Comment (albertshau) [inline src/main/java/co/cask/cdc/plugins/common/References.java]: update copyright year

File: src/main/java/co/cask/cdc/plugins/common/CDCReferencePluginConfig.java [hunk 8]
```diff
@@ -0,0 +1,22 @@
+package co.cask.cdc.plugins.common;
```
Comment (albertshau) [inline src/main/java/co/cask/cdc/plugins/common/CDCReferencePluginConfig.java:17]: missing copyright

File: src/test/java/co/cask/cdc/plugins/sink/CDCBigTableConfigTest.java [hunk 9]
```diff
@@ -0,0 +1,134 @@
...
+  @Test
+  public void testValidateValidConfig() {
+    CDCBigTableConfig config = new CDCBigTableConfig(
+      VALID_REF,
+      VALID_INSTANCE,
+      VALID_PROJECT,
+      VALID_ACCOUNT_FILE_PATH
+    );
+
+    Assertions.assertThatCode(config::validate)
```
Comment (albertshau) [inline src/test/java/co/cask/cdc/plugins/sink/CDCBigTableConfigTest.java]: This can be replaced with just a call to config.validate().

File: src/test/java/co/cask/cdc/plugins/sink/CDCBigTableConfigTest.java [hunk 10]
```diff
@@ -0,0 +1,134 @@
...
+  @Test
+  public void testValidateReference() {
+    CDCBigTableConfig config = new CDCBigTableConfig(
+      "",
+      VALID_INSTANCE,
+      VALID_PROJECT,
+      VALID_ACCOUNT_FILE_PATH
+    );
+
+    Assertions.assertThatThrownBy(config::validate)
```
Comment (albertshau) [inline src/test/java/co/cask/cdc/plugins/sink/CDCBigTableConfigTest.java]: this should be replaced with a call to config.validate(), where you catch the exception and check that e.getProperty() is correct.

Comment (albertshau) [reply src/test/java/co/cask/cdc/plugins/sink/CDCBigTableConfigTest.java]: same for the next few tests

File: src/test/java/co/cask/cdc/plugins/sink/CDCBigTableConfigTest.java [hunk 11]
```diff
@@ -0,0 +1,134 @@
...
+  public void testResolveProjectId() {
+    CDCBigTableConfig config = new CDCBigTableConfig(
+      VALID_REF,
+      VALID_INSTANCE,
+      null,
+      VALID_ACCOUNT_FILE_PATH
+    );
+
+    Assertions.assertThat(config.resolveProject())
+      .isEqualTo(ServiceOptions.getDefaultProjectId());
```
Comment (albertshau) [inline src/test/java/co/cask/cdc/plugins/sink/CDCBigTableConfigTest.java]: Tests should use plain junit assertions when possible. Does this give you anything more than normal junit Assert.assertEquals()?

File: src/test/java/co/cask/cdc/plugins/sink/CDCBigTableConfigTest.java [hunk 12]
```diff
@@ -0,0 +1,134 @@
...
+  @Test
+  public void testServiceAccountFilePath() {
+    CDCBigTableConfig config = new CDCBigTableConfig(
+      VALID_REF,
+      VALID_INSTANCE,
+      VALID_PROJECT,
+      null
+    );
+
+    Assertions.assertThat(config.resolveServiceAccountFilePath())
```
Comment (albertshau) [inline src/test/java/co/cask/cdc/plugins/sink/CDCBigTableConfigTest.java]: Similar question here

File: src/test/java/co/cask/cdc/plugins/sink/CDCBigTableConfigTest.java [hunk 13]
```diff
@@ -0,0 +1,134 @@
+package co.cask.cdc.plugins.sink;
```
Comment (albertshau) [inline src/test/java/co/cask/cdc/plugins/sink/CDCBigTableConfigTest.java:17]: missing copyright

**Review by albertshau — COMMENTED** [review]
just noticed that a lot of files are missing copyright. This is supposed to be in the rat check, have we not set it up for this project?

File: src/test/java/co/cask/cdc/plugins/sink/CDCBigTableConfigTest.java (hunk 10 above)
Comment (mlozbin-cybervisiontech) [reply src/test/java/co/cask/cdc/plugins/sink/CDCBigTableConfigTest.java]: > just noticed that a lot of files are missing copyright. This is supposed to be in the rat check, have we not set it up for this project?

This plugin was not configured for this project as well as for google-cloud. I may submit another PR with rat plugin and updated copyrights in all classes.

File: src/test/java/co/cask/cdc/plugins/sink/CDCBigTableConfigTest.java (hunk 11 above)
Comment (mlozbin-cybervisiontech) [reply src/test/java/co/cask/cdc/plugins/sink/CDCBigTableConfigTest.java]: > just noticed that a lot of files are missing copyright. This is supposed to be in the rat check, have we not set it up for this project?

This plugin was not configured for this project as well as for google-cloud. I may submit another PR with rat plugin and updated copyrights in all classes.

**Comment by mlozbin-cybervisiontech** [comment]
> just noticed that a lot of files are missing copyright. This is supposed to be in the rat check, have we not set it up for this project?

@albertshau This plugin was not configured for this project as well as for google-cloud. I may submit another PR with rat plugin and updated copyrights in all classes.

File: src/main/java/co/cask/cdc/plugins/common/References.java (hunk 7 above)
Comment (mlozbin-cybervisiontech) [reply src/main/java/co/cask/cdc/plugins/common/References.java]: removed this class and avoided duplication of code from IdUtils

File: src/main/java/co/cask/cdc/plugins/common/CDCReferencePluginConfig.java (hunk 6 above)
Comment (mlozbin-cybervisiontech) [reply src/main/java/co/cask/cdc/plugins/common/CDCReferencePluginConfig.java]: removed References class and avoided duplication of code from IdUtils

File: src/main/java/co/cask/cdc/plugins/common/CDCReferencePluginConfig.java (hunk 8 above)
Comment (mlozbin-cybervisiontech) [reply src/main/java/co/cask/cdc/plugins/common/CDCReferencePluginConfig.java:17]: fixed

File: src/test/java/co/cask/cdc/plugins/sink/CDCBigTableConfigTest.java (hunk 9 above)
Comment (mlozbin-cybervisiontech) [reply src/test/java/co/cask/cdc/plugins/sink/CDCBigTableConfigTest.java]: fixed

PS: I prefer BDD in cases where we want to make an accent on what is actually tested here. I understand your point and will simplify such tests in future.

File: src/test/java/co/cask/cdc/plugins/sink/CDCBigTableConfigTest.java (hunk 10 above)
Comment (mlozbin-cybervisiontech) [reply src/test/java/co/cask/cdc/plugins/sink/CDCBigTableConfigTest.java]: Added invalid property check. Can we use AssertJ here to have much more readable assertion and failure message? You can compare it with a code to make similar thing https://github.com/joel-costigliola/assertj-examples/blob/master/assertions-examples/src/test/java/org/assertj/examples/FailUsageExamples.java.
AssertJ also allows to easily extract test conditions to test framework to make tests for all plugins look similar and more readable. Currently there are no guides and multiple approaches.

File: src/test/java/co/cask/cdc/plugins/sink/CDCBigTableConfigTest.java (hunk 11 above)
Comment (mlozbin-cybervisiontech) [reply src/test/java/co/cask/cdc/plugins/sink/CDCBigTableConfigTest.java]: Actually assertEquals is almost the only case where JUnit assertion works in same way as AssertJ one. And even in this case there are some differences:
- JUnit assertions provoke developer to pass "expected" and "actual" params in wrong order and you'll have confusing failure message. This is fine when test failure is reproducible but often leads to loss of time when bug is reproducible only on CI or once per week (for example).
- JUnit's assertEquals does not have custom representation, custom comparator and soft assertions support.
- JUnit's assertEquals is less smart than one from AssertJ in some cases. For example: if you compare numbers with floating point assertJ will use comparison delta by default.

I also wanted to add why I prefer AssertJ in general:
- it provides much better failure messages. For example with JUnit:
    - you often use assertNotNull that will fail with AssertionError with no message (no details at all)
    - with code like Assert.assertTrue(one.getSchema().getFields().size() == two.getSchema().getFields().size()) you will not see wrong actual schema to analyze what happened. This is a problem when bug is not reproducible.
- it is single library that covers all possible cases where you previously needed Junit, TestNG or Hamcrest assertion libraries.
- it makes code much more readable and concise.
- it provides custom assertions functionality and other extendability options that allows to make testing frameworks. That approach reduces code duplication and makes all tests look similar without complex guidelines.
- it gives great auto-completion for IDE which is very handy with custom assertions.
- more detailed key difference overview: https://professional-practical-programmer.com/2016/06/26/assert-j/
- AsserJ features highlight: http://joel-costigliola.github.io/assertj/assertj-core-features-highlight.html

In summary. There are definitely some cases where AssertJ does better job than Junit assertions. And in this case there is a choice: to use 3 libraries depending on use case (Junit, hamcrest, AssertJ) or to use only 1 library in all cases (AssertJ). As for me, second approach is easier to describe in guides and to understand than trying to describe what library should be used in what case.

Please, consider this as my personal opinion and just a proposal. Thank you for your time.

File: src/test/java/co/cask/cdc/plugins/sink/CDCBigTableConfigTest.java (hunk 13 above)
Comment (mlozbin-cybervisiontech) [reply src/test/java/co/cask/cdc/plugins/sink/CDCBigTableConfigTest.java:17]: fixed

File: src/test/java/co/cask/cdc/plugins/sink/CDCBigTableConfigTest.java (hunk 10 above)
Comment (albertshau) [reply src/test/java/co/cask/cdc/plugins/sink/CDCBigTableConfigTest.java]: In general I don't think it's great to assert exception messages in tests as it is very brittle and tends to just test how things happen to be implemented rather than testing an explicit contract. 

That is why I suggested catching the InvalidConfigPropertyException and asserting that the property name is correct. That is an actual contract that the plugin must follow.

For other scenarios it may make sense to do what you suggested.

File: src/test/java/co/cask/cdc/plugins/sink/CDCBigTableConfigTest.java (hunk 11 above)
Comment (albertshau) [reply src/test/java/co/cask/cdc/plugins/sink/CDCBigTableConfigTest.java]: I agree that junit doesn't have the most features, but I think other more powerful testing frameworks can sometimes allow you to get away with not having the best abstractions in the actual code. So I hesitate to add it where it is not necessary. As a general rule, I think it makes sense to start with only JUnit, since that is the standard across the other repositories, and only add others if there is a compelling reason to do it. In this particular case, it doesn't seem to give us anything that you couldn't do in junit.

File: src/test/java/co/cask/cdc/plugins/sink/CDCBigTableConfigTest.java [hunk 14]
```diff
@@ -36,7 +55,7 @@ public void testValidateReference() {
     );
 
     Assertions.assertThatThrownBy(config::validate)
-      .isInstanceOf(InvalidConfigPropertyException.class);
+      .isInstanceOfSatisfying(InvalidConfigPropertyException.class, withProperty(Constants.Reference.REFERENCE_NAME));
```
Comment (albertshau) [inline src/test/java/co/cask/cdc/plugins/sink/CDCBigTableConfigTest.java]: It's simpler to just do:

```
try {
  config.validate()
} catch (InvalidConfigPropertyException e) {
  Assert.assertEquals(Constants.Reference.REFERENCE_NAME, e.getProperty());
}
```

this is just normal java that mirrors how the caller is expected to behave.

File: src/test/java/co/cask/cdc/plugins/sink/CDCBigTableConfigTest.java (hunk 10 above)
Comment (mlozbin-cybervisiontech) [reply src/test/java/co/cask/cdc/plugins/sink/CDCBigTableConfigTest.java]: I was trying to say about failure messages in test logs/reports, not about asserting exception messages

File: src/test/java/co/cask/cdc/plugins/sink/CDCBigTableConfigTest.java (hunk 11 above)
Comment (mlozbin-cybervisiontech) [reply src/test/java/co/cask/cdc/plugins/sink/CDCBigTableConfigTest.java]: fixed

File: src/test/java/co/cask/cdc/plugins/sink/CDCBigTableConfigTest.java (hunk 12 above)
Comment (mlozbin-cybervisiontech) [reply src/test/java/co/cask/cdc/plugins/sink/CDCBigTableConfigTest.java]: fixed

File: src/test/java/co/cask/cdc/plugins/sink/CDCBigTableConfigTest.java (hunk 14 above)
Comment (mlozbin-cybervisiontech) [reply src/test/java/co/cask/cdc/plugins/sink/CDCBigTableConfigTest.java]: fixed

**Comment by mlozbin-cybervisiontech** [comment]
@albertshau, addressed all recent comments.

**Comment by mlozbin-cybervisiontech** [comment]
> > just noticed that a lot of files are missing copyright. This is supposed to be in the rat check, have we not set it up for this project?
> 
> @albertshau This plugin was not configured for this project as well as for google-cloud. I may submit another PR with rat plugin and updated copyrights in all classes.

created https://github.com/data-integrations/change-data-capture/pull/9

**Review by albertshau — APPROVED** [review]
lgtm
